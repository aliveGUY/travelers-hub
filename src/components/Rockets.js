import React from 'react';
import {
  Container, Col, Row, Button,
} from 'react-bootstrap';
import { useSelector } from 'react-redux/es/exports';

export default function Rockets() {
  const rockets = useSelector((state) => state.rockets);
  console.log(rockets);
  return (
    <Container>
      {
        rockets.map((rocket) => (
          <Row key={rocket.id} className="mb-3">
            <Col sm={4}>
              <img src={rocket.images[0]} alt="Rocket" />
            </Col>
            <Col sm={8}>
              <p className="h4">{rocket.name}</p>
              <p>{rocket.description}</p>
              <Button variant="primary" size="md">Reserve Rocket</Button>
            </Col>
          </Row>
        ))
      }
    </Container>
  );
}
