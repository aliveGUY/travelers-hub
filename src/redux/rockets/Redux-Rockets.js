import axios from 'axios';
// Actions
const FETCHED_DATA = 'FETCHED_DATA';

// Reducer
export default function reducer(state = {}, action = {}) {
  switch (action.type) {
    case FETCHED_DATA:
      return action.list;
    default: return state;
  }
}

// Action Creators
// thunks
export const dispatchRockets = () => async (dispatch) => {
  const response = await axios('https://api.spacexdata.com/v3/rockets');
  const list = response.data.map((rocket) => ({
    id: rocket.id,
    name: rocket.rocket_name,
    type: rocket.rocket_type,
    images: rocket.flickr_images,
  }));
  dispatch({ type: FETCHED_DATA, list });
};
