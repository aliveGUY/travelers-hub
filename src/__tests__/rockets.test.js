import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {
  render, fireEvent, screen,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import Rockets from '../components/Rockets';
import { reserveRocket } from '../redux/rockets/Redux-Rockets';

const mockStore = configureStore([]);

describe('Rockets component test', () => {
  const rockets = [
    {
      id: '0',
      name: 'Rocket 0',
      description: 'Rocket description',
      images: ['img 0', 'img 1'],
      reserved: false,
    },
  ];

  test('Component renders/ Snapshot test', () => {
    const store = mockStore({
      rockets,
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <Rockets />
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Reserve rocket functionality', () => {
    const store = mockStore({
      rockets,
    });
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    fireEvent.click(screen.getByText('Reserve Rocket'));
    const firedAction = store.getActions();

    expect(...firedAction).toEqual(reserveRocket('0'));
  });

  test('Cancel rocket reservation functionality', () => {
    const store = mockStore({
      rockets: [
        {
          id: '0',
          name: 'Rocket 0',
          description: 'Rocket description',
          images: ['img 0', 'img 1'],
          reserved: true,
        },
      ],
    });
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    fireEvent.click(screen.getByText('Cancel Reservation Rocket'));
    const firedAction = store.getActions();

    expect(...firedAction).toEqual(reserveRocket('0'));
  });
});
