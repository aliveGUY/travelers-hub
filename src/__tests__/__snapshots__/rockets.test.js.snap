// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets component test Component renders/ Snapshot test 1`] = `
<div
  className="container"
>
  <div
    className="m-3 row"
  >
    <div
      className="col-sm-4"
    >
      <img
        alt="Rocket"
        src="img 0"
      />
    </div>
    <div
      className="col-sm-8"
    >
      <p
        className="h4"
      >
        Rocket 0
      </p>
      <p
        className="d-inline"
      >
        Rocket description
      </p>
      <br />
      <button
        className="mt-3 btn btn-primary btn-md"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</div>
`;
